packages/plugins/i18n/admin/src/components/CMHeaderActions.tsx

```diff
@@ -226,6 +237,299 @@ const DeleteLocaleAction: DocumentActionComponent = ({
   };
 };
 
+/* -------------------------------------------------------------------------------------------------
+ * BulkPublishAction
+ * -----------------------------------------------------------------------------------------------*/
+
+export type LocaleStatus = {
+  locale: string;
+  status: Modules.Documents.Params.PublicationStatus.Kind | 'modified';
+};
+
+const BulkLocalePublishAction: DocumentActionComponent = ({
+  document: baseDocument,
+  documentId,
+  model,
+  collectionType,
+}) => {
+  const baseLocale = baseDocument?.locale ?? null;
+
+  const [{ query }] = useQueryParams<{ status: 'draft' | 'published' }>();
+
+  const params = React.useMemo(() => buildValidParams(query), [query]);
+  const isPublishedTab = query.status === 'published';
+
+  const { formatMessage } = useIntl();
+  const { hasI18n, canPublish } = useI18n();
+  const { toggleNotification } = useNotification();
+  const { _unstableFormatAPIError: formatAPIError } = useAPIErrorHandler();
+
+  const [selectedRows, setSelectedRows] = React.useState<any[]>([]);
+  const [isConfirmationOpen, setIsConfirmationOpen] = React.useState<boolean>(false);
+
+  const { publishMany: publishManyAction } = useDocumentActions();
+  const {
+    document,
+    meta: documentMeta,
+    schema,
+    validate,
+  } = useDocument({
+    model,
+    collectionType,
+    documentId,
+    params: {
+      locale: baseLocale,
+    },
+  });
+
+  const { data: localesMetadata = [] } = useGetLocalesQuery();
+
+  const headers = [
+    {
+      label: formatMessage({
+        id: 'global.name',
+        defaultMessage: 'Name',
+      }),
+      name: 'name',
+    },
+    {
+      label: formatMessage({
+        id: getTranslation('CMEditViewBulkLocale.status'),
+        defaultMessage: 'Status',
+      }),
+      name: 'status',
+    },
+    {
+      label: formatMessage({
+        id: getTranslation('CMEditViewBulkLocale.publication-status'),
+        defaultMessage: 'Publication Status',
+      }),
+      name: 'publication-status',
+    },
+  ];
+
+  const [rows, validationErrors] = React.useMemo(() => {
```

Comment:
i find this whole block really hard to follow, can we add some comments to explain why we're doing things?